pkg/ingester/ingester.go

```diff
@@ -1289,6 +1324,15 @@ func (i *Ingester) pushSamplesToAppender(userID string, timeseries []mimirpb.Pre
 	handleAppendError := func(err error, timestamp int64, labels []mimirpb.LabelAdapter) bool {
 		stats.failedSamplesCount++
 
+		// Even though IsSoftAppendError checks the same errors as the below switch case, keeping all the error checks
```

Comment:
Less concerned with the added latency, but this is dying to be pulled apart to fix the duplication. I think there's a solution where the switch/case is moved to a new type and it either does or does not invoke callbacks based on initialization.

Ingester would construct this type so that the callbacks are active; block-builder would construct the type so that they're no-ops.
Then each would call into `softErr: = ingesterr.ProcessErr(err)`

ProcessErr would look like this:

```
func (e *errorProcessor) ProcessErr(err error) bool {
   switch {
		case errors.Is(err, storage.ErrOutOfBounds):
         e.outOfBounds(labels)
			return true
      case errors.Is(err, storage.ErrOutOfOrderSample):
         e.oooSample(labels)
			return true
       ...
   }
   return false
}
```

(Or, similarly, a map of {errtype -> callback func} could be provided.) What do you think?

Reply:
This looks like a good idea, but I am concerned about allocs for ingesters this way on the hot path. `updateFirstPartial()` is a request specific thing. So we will need a new `errorProcessor{}` everytime. Might lead to more allocs or probably not. I will try it and benchmark.

Reply:
Probably no extra type necessary, just a separate function. But yeah whether the per-request callbacks are stuffed in a map or a struct, they might escape to the heap.

Reply:
Looks like it didn't add any overhead. Made the change. Should be easier to keep things updated now!